src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -364,32 +378,34 @@ protected virtual TypeSymbolWithAnnotations SubstituteType(AbstractTypeMap typeM
             var newTypeWithModifiers = typeMap.SubstituteType(this.TypeSymbol, withTupleUnification);
             bool newIsAnnotated = this.IsAnnotated || newTypeWithModifiers.IsAnnotated;
 
-            if (!TypeSymbolEquals(newTypeWithModifiers, TypeCompareKind.CompareNullableModifiersForReferenceTypes) ||
-                !newTypeWithModifiers.CustomModifiers.IsEmpty ||
-                newIsAnnotated != this.IsAnnotated ||
-                newCustomModifiers != this.CustomModifiers)
+            // PROTOTYPE(NullableReferenceTypes): Can we use Equals instead?
+            if (TypeSymbolEquals(newTypeWithModifiers, TypeCompareKind.CompareNullableModifiersForReferenceTypes) &&
+                newTypeWithModifiers.CustomModifiers.IsEmpty &&
+                newIsAnnotated == this.IsAnnotated &&
+                newCustomModifiers == this.CustomModifiers)
             {
-                var newIsNullableType = newTypeWithModifiers.TypeSymbol.IsNullableType();
-                if (newIsNullableType || !newIsAnnotated)
-                {
-                    if (newCustomModifiers.IsEmpty)
-                    {
-                        return newTypeWithModifiers;
-                    }
-                    newIsAnnotated = newTypeWithModifiers.IsAnnotated;
-                }
-                else
+                // PROTOTYPE(NullableReferenceTypes): We're dropping newTypeWithModifiers.NonNullTypes!
+                return this; // substitution had no effect on the type or modifiers
+            }
+
+            bool newIsNullableType = newTypeWithModifiers.TypeSymbol.IsNullableType();
+            if (newIsNullableType || !newIsAnnotated)
```

Comment:
I'm still having trouble following this condition and first alternative. I can see how we'd want to fix `IsAnnotated` when new is a nullable type. But I don't understand the other case (`!newIsAnnotated`). Could you clarify?

Reply:
I think you're right, there is no such case. I've removed that clause.

---
In reply to: [203896423](https://github.com/dotnet/roslyn/pull/28687#discussion_r203896423) [](ancestors = 203896423)